### PR TITLE
Support Weight Normalization in fundamental layers

### PR DESCRIPTION
**Pull request by henry0312**
[Weight Normalization](https://arxiv.org/abs/1602.07868 "Weight Normalization") is a new way to accelerate training of Deep Neural Networks, which is accepted in NIPS 2016.
cf. http://papers.nips.cc/paper/6113-weight-normalization-a-simple-reparameterization-to-accelerate-training-of-deep-neural-networks

I added weight normalization to some fundamental layers:

* Core
  * Dense
* Convolutional
  * Convolution1D
  * Convolution2D
  * Deconvolution2D
* Recurrent
  * SimpleRNN
  * GRU
  * LSTM

### CODE REVIEW

**Comment by fchollet** [comment]
Weight normalization has been in use for years. In Keras, you can use it by setting the `W_constraint` argument (available in most layers that have weights) to `unitnorm`. See the "constraints" section in the docs.

**Comment by henry0312** [comment]
@fchollet `W_constraint` doesn't cover all of the Weight Normalization that is accepted NIPS 2016

**Comment by henry0312** [comment]
For example, please see https://github.com/fchollet/keras/pull/4436/files#diff-05ec8fc63f0e9bb35f99c0ca76a66d0eR703.

**Comment by fchollet** [comment]
FYI there is no difference between backpropping through W vs. backpropping through W * g (both trainable). It's the same linear projection in both cases. Sure, now the normalization only uses the norm of W, so that's a difference. Albeit a minor one. Unclear that is it at all necessary or that the benefits aren't coming from the added capacity.

We are not making API changes to core layers to accommodate random new techniques such as adding a scaling matrix to weight matrices...

**Comment by oleg-trott** [comment]
In the special case of the usual feed-forward NNs, no biases and LeakyReLU or ReLU nonlinearities, the scalar factors can all be absorbed into the last layer's weights (not subject to WN).

However, in general, *e.g.* with sigmoid nonlinearities, or ResNets, this is no longer the case.

**Comment by oleg-trott** [comment]
Reading the paper more carefully, I think it's implied that `g` is a different scalar for each channel, so we can't actually absorb them into the last layer without losing generality.

ResNets seem like an obvious example of where using `unitnorm` without the scaling factors would be deleterious, although I haven't tried it.

**Comment by redst4r** [comment]
also note that there's reference [implementation](https://github.com/openai/weightnorm) where the weight-normalization is just contained in the optimizer, no need to change the layer implementation
